common/src/main/java/io/druid/math/expr/ExprEval.java

```diff
@@ -0,0 +1,267 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.math.expr;
+
+import com.google.common.base.Strings;
+import io.druid.java.util.common.IAE;
+
+/**
+ */
+public abstract class ExprEval<T>
+{
+  public static ExprEval of(long longValue)
+  {
+    return new LongExprEval(longValue);
+  }
+
+  public static ExprEval of(double longValue)
+  {
+    return new DoubleExprEval(longValue);
+  }
+
+  public static ExprEval of(String stringValue)
+  {
+    return new StringExprEval(stringValue);
+  }
+
+  public static ExprEval of(boolean value, ExprType type)
+  {
+    switch (type) {
+      case DOUBLE:
+        return ExprEval.of(value ? 1D : 0D);
+      case LONG:
+        return ExprEval.of(value ? 1L : 0L);
+      case STRING:
+        return ExprEval.of(String.valueOf(value));
+      default:
+        throw new IllegalArgumentException("invalid type " + type);
+    }
+  }
+
+  public static ExprEval bestEffortOf(Object val)
+  {
+    if (val instanceof ExprEval) {
+      return (ExprEval) val;
+    }
+    if (val instanceof Number) {
+      if (val instanceof Float || val instanceof Double) {
+        return new DoubleExprEval((Number)val);
+      }
+      return new LongExprEval((Number)val);
+    }
+    return new StringExprEval(val == null ? null : String.valueOf(val));
```

Comment:
Hmm, I think this means all nulls, even nulls of columns we want to treat as numbers (e.g. missing columns), are getting represented as StringExprEvals. I guess that works but seems strange. It might make sense for null to have its own datatype.

I guess I don't have strong feelings on this now for this PR but we should consider null handling before finalizing the expressions API.